rueidisprob/bloomfilter.go

```diff
@@ -0,0 +1,344 @@
+// BloomFilter uses 128-bit murmur3 hash function.
+type BloomFilter interface {
+	// Add adds an item to the Bloom filter.
+	Add(ctx context.Context, key string) error
+
+	// AddMulti adds one or more items to the Bloom filter.
+	// NOTE: If keys are too many, it can block the Redis server for a long time.
+	AddMulti(ctx context.Context, keys []string) error
+
+	// Exists checks if an item is in the Bloom filter.
+	Exists(ctx context.Context, key string) (bool, error)
+
+	// ExistsMulti checks if one or more items are in the Bloom filter.
+	// Returns a slice of bool values where each bool indicates whether the corresponding key was found.
+	ExistsMulti(ctx context.Context, keys []string) ([]bool, error)
+
+	// Reset resets the Bloom filter.
+	Reset(ctx context.Context) error
+
+	// Delete deletes the Bloom filter.
+	Delete(ctx context.Context) error
+
+	// Count returns count of items in Bloom filter.
+	Count(ctx context.Context) (uint, error)
+}
+
+type bloomFilter struct {
+	client rueidis.Client
+
+	// name is the name of the Bloom filter.
+	// It is used as a key in the Redis.
+	name string
+
+	// counter is the name of the counter.
+	counter string
+
+	// hashIterations is the number of hash functions to use.
+	hashIterations      uint
+	hashIterationString string
+
+	// size is the number of bits to use.
+	size uint
+
+	addMultiScript *rueidis.Lua
+	addMultiKeys   []string
+
+	existsMultiScript *rueidis.Lua
+	existsMultiKeys   []string
+}
+
+// NewBloomFilter creates a new Bloom filter.
+// NOTE: 'name:c' is used as a counter key in the Redis
+// to keep track of the number of items in the Bloom filter for Count method.
+func NewBloomFilter(
+	client rueidis.Client,
+	name string,
+	expectedNumberOfItems uint,
+	falsePositiveRate float64,
+) (BloomFilter, error) {
+	if len(name) == 0 {
+		return nil, ErrEmptyName
+	}
+
+	if falsePositiveRate < 0 {
+		return nil, ErrFalsePositiveRateNegative
+	}
+	if falsePositiveRate > 1 {
+		return nil, ErrFalsePositiveRateGreaterThanOne
+	}
+
+	size := numberOfBits(expectedNumberOfItems, falsePositiveRate)
+	if size == 0 {
+		return nil, ErrBitsSizeZero
+	}
+	if size > maxSize {
+		return nil, ErrBitsSizeTooLarge
+	}
+	hashIterations := numberOfHashFunctions(size, expectedNumberOfItems)
+
+	// NOTE: https://redis.io/docs/reference/cluster-spec/#hash-tags
+	bfName := "{" + name + "}"
+	counterName := bfName + ":c"
+	return &bloomFilter{
+		client:              client,
+		name:                bfName,
+		counter:             counterName,
+		hashIterations:      hashIterations,
+		hashIterationString: strconv.FormatUint(uint64(hashIterations), 10),
+		size:                size,
+		addMultiScript:      rueidis.NewLuaScript(addMultiScript),
+		addMultiKeys:        []string{bfName, counterName},
+		existsMultiScript:   rueidis.NewLuaScript(existsMultiScript),
+		existsMultiKeys:     []string{bfName},
+	}, nil
+}
+
+func numberOfBits(n uint, r float64) uint {
+	if r == 0 {
+		r = math.SmallestNonzeroFloat64
```

Comment:
Why don't we just reject any r which is smaller than math.SmallestNonzeroFloat64?

Reply:
[e7cd23c](https://github.com/redis/rueidis/pull/493/commits/e7cd23c9e2374c581482087955d2c9ac19c8b7b9)